File: algorithms/swift/search/BinarySearch.swift
import Foundation

func binarySearch<T: Comparable>(_ array: [T], key: T) -> Int? {
    var lowerBound = 0
    var upperBound = array.count
    
    while lowerBound < upperBound {
        let midIndex = lowerBound + (upperBound - lowerBound) / 2
        if array[midIndex] == key {
            return midIndex
        } else if array[midIndex] < key {
            lowerBound = midIndex + 1
        } else {
            upperBound = midIndex
        }
    }
    return nil
}

let numbers = [2, 3, 4, 10, 40]
if let index = binarySearch(numbers, key: 10) {
    print("Element found at index: \(index)")
} else {
    print("Element not found")
}

File: algorithms/dart/search/binary_search.dart
int binarySearch(List<int> list, int target) {
  int min = 0;
  int max = list.length - 1;

  while (min <= max) {
    int mid = ((min + max) / 2).floor();
    if (target == list[mid]) {
      return mid;
    } else if (target < list[mid]) {
      max = mid - 1;
    } else {
      min = mid + 1;
    }
  }
  return -1;
}

void main() {
  List<int> list = [2, 3, 4, 10, 40];
  int target = 10;
  
  int index = binarySearch(list, target);
  
  if (index != -1) {
    print('Element found at index: $index');
  } else {
    print('Element not found');
  }
}
